### PR TITLE
[Merged by Bors] - feat: Positivity extension for `Finset.sum`

### PR DESCRIPTION
**Pull request by YaelDillies**
Also define a new `aesop` rule-set and an auxiliary metaprogram `proveFinsetNonempty` for dealing with `Finset.Nonempty` conditions.

From LeanAPAP

Co-authored-by: Alex J. Best <alex.j.best@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 1]
```diff
@@ -816,3 +816,61 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+TODO: Replace by an aesop ruleset? -/
+def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
+    MetaM Q(Finset.Nonempty $s) := do
+  try
+    let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) | failure
+    pure $ .fvar fv
+  catch _ => do
+    let ~q(@univ _ $fi) := s | failure
+    let _no ← synthInstanceQ q(Nonempty $α)
+    pure $ q(Finset.univ_nonempty (α := $α))
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:839]:
```suggestion
    pure <| q(Finset.univ_nonempty (α := $α))
```
Please try to stick to the style guide, though I realize muscle memory can be hard to banish

**Comment by eric-wieser** [comment]
I think you've marked the dependency backwards?

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 2]
```diff
@@ -816,3 +816,61 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+    -- Try to show that the sum is nonnegative
+    catch _ => do
+      let pbody ← rbody.toNonneg
+      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pbody
+      let pα' ← synthInstanceQ q(OrderedAddCommMonoid $α)
+      assertInstancesCommute
+      let proof := q(@sum_nonneg $ι $α $pα' $f $s fun i _ ↦ $pr i)
+      proof.check
+      return .nonnegative proof
+  | _ => throwError "not Finset.sum"
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:874]: I think this is better without the `try`/`catch`:
```lean
def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
    MetaM (Option Q(Finset.Nonempty $s)) := do
  if let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) then
    return some (.fvar fv)
  else if let ~q(@univ _ $fi) := s then
    if let .some _no ← trySynthInstanceQ q(Nonempty $α) then
      return some q(Finset.univ_nonempty)
    else
      return none
  else
    return none

/-- The `positivity` extension which proves that `∑ i in s, f i` is nonnegative if `f` is, and
positive if each `f i` is and `s` is nonempty.

TODO: The following example does not work
```
example (s : Finset ℕ) (f : ℕ → ℤ) (hf : ∀ n, 0 ≤ f n) : 0 ≤ s.sum f := by positivity
```
because `compareHyp` can't look for assumptions behind binders.
-/
@[positivity Finset.sum _ _]
def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
  match e with
  | ~q(@Finset.sum _ $ι $instα $s $f) =>
    let i : Q($ι) ← mkFreshExprMVarQ q($ι) .syntheticOpaque
    have body : Q($α) := Expr.betaRev f #[i]
    let rbody ← core zα pα body
    -- Try to show that the sum is positive
    let p_pos : Option Q(0 < $e) := ← (do
      let .positive pbody := rbody | pure none -- Fail if the body is not provably positive
      let .some ps ← proveFinsetNonempty s | pure none
      let .some pα' ← trySynthInstanceQ q(OrderedCancelAddCommMonoid $α) | pure none
      assertInstancesCommute
      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pbody
      return some q(@sum_pos $ι $α $pα' $f $s (fun i _ ↦ $pr i) $ps))
    if let some p_pos := p_pos then
      return .positive p_pos
    else
      let pbody ← rbody.toNonneg
      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pbody
      let pα' ← synthInstanceQ q(OrderedAddCommMonoid $α)
      assertInstancesCommute
      let proof := q(@sum_nonneg $ι $α $pα' $f $s fun i _ ↦ $pr i)
      proof.check
      return .nonnegative proof
  | _ => throwError "not Finset.sum"
  ```

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Order.lean:874]: This has the advantage of not catching anything that you don't expect; for instance, if `mkLambdaFVars` fails somehow, or `(try)SynthInstanceQ` fails rather than finds nothing.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 3]
```diff
@@ -816,3 +816,61 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+positive if each `f i` is and `s` is nonempty.
+
+TODO: The following example does not work
+```
+example (s : Finset ℕ) (f : ℕ → ℤ) (hf : ∀ n, 0 ≤ f n) : 0 ≤ s.sum f := by positivity
+```
+because `compareHyp` can't look for assumptions behind binders.
+-/
+@[positivity Finset.sum _ _]
+def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Order.lean:851]: Does this PR have the wrong title? This isn't `Finset.prod`

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:851]: Yes sorry, I accidentally swapped the content of both PRs. Feel free to fix it, I'm off to bed.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 4]
```diff
@@ -816,3 +816,61 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+open Qq Lean Meta Finset
+
+/-- Prove that a finset is nonempty from a local assumption or from the fact that the finset is
+`univ` in a nonempty fintype.
+
+Note that this does not do any complicated reasoning: the only ways it can prove `s` is nonempty is
+if there is a local `s.Nonempty` hypothesis or `s = (univ : Finset α)` and `Nonempty α` can be
+synthesized by TC inference.
+
+TODO: Replace by an aesop ruleset? -/
```
Comment (Vierkantor) [inline Mathlib/Algebra/BigOperators/Order.lean:830]: I tried for a bit to get this to work, but no success. Here's a Zulip thread: https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/Calling.20aesop.20from.20a.20tactic.20to.20obtain.20an.20.60Expr.60

**Review by Vierkantor — APPROVED** [review]
This is as good as I can make it right now. I would like to wait on the Zulip thread about calling into aesop, but if that doesn't resolve with a positive answer in a few days I think we can go ahead and merge.

File: Mathlib/Algebra/BigOperators/Order.lean [hunk 5]
```diff
@@ -816,3 +816,61 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
...
+def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
+    MetaM (Option Q(Finset.Nonempty $s)) := do
+  if let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) then
+    return some (.fvar fv)
+  else if let ~q(@univ _ $fi) := s then
+    if let .some _no ← trySynthInstanceQ q(Nonempty $α) then
+      return some q(Finset.univ_nonempty)
+    else return none
+  else return none
```
Comment (Vierkantor) [inline Mathlib/Algebra/BigOperators/Order.lean:839]: Thanks to Jannis' help, I came up with:
```suggestion
-/
def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
    MetaM (Option Q(Finset.Nonempty $s)) := do
  -- Aesop expects to operate on goals, so we're going to make a new goal.
  let goal ← Lean.Meta.mkFreshExprMVar q(Finset.Nonempty $s)
  let mvar := goal.mvarId!
  -- We want this to be fast, so use only the basic and `Finset.Nonempty` specific rules.
  let rulesets ← Aesop.Frontend.getGlobalRuleSets #[`builtin, `Finset.Nonempty]
  -- Fail if the new goal is not closed.
  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
  let (remainingGoals, _) ← Aesop.search mvar (.some rules)
  -- Fail if there are open goals remaining, this serves as an extra check for the
  -- Aesop configuration option `terminal := true`.
  if remainingGoals.size > 0 then return none
  Lean.getExprMVarAssignment? mvar
```

With the following extra declarations:
```lean
-- `Finset.Nonempty` rules try to prove that a given finset is nonempty,
-- for use in the `Finset.sum` positivity extension.
declare_aesop_rule_sets [Finset.Nonempty] (default := true)
```
and in a separate file:
```lean
attribute [aesop unsafe apply (rule_sets [Finset.Nonempty])] Finset.univ_nonempty
```
I would argue that the `declare_aesop_rule_sets` line should go into a new file (maybe `Mathlib/Data/Finset/Attr.lean`?) and the `attribute` can directly go onto the declaration of `Finset.univ_nonempty`.

Comment (Vierkantor) [reply Mathlib/Algebra/BigOperators/Order.lean:839]: (It is still an annoyance that environment extensions only become visible after importing!)

File: test/positivity.lean [hunk 6]
```diff
@@ -322,6 +323,24 @@ example {r : ℝ≥0} (hr : 0 < r) : (0 : ℝ) < r := by positivity
...
+example (n : ℕ) (a : ℕ → ℤ) : 0 ≤ ∑ j in range n, a j^2 := by positivity
+example (a : ULift.{2} ℕ → ℤ) (s : Finset (ULift.{2} ℕ)) : 0 ≤ ∑ j in s, a j ^ 2 := by positivity
+example (n : ℕ) (a : ℕ → ℤ) : 0 ≤ ∑ j : Fin 8, ∑ i in range n, (a j ^ 2 + i ^ 2) := by positivity
+example (n : ℕ) (a : ℕ → ℤ) : 0 < ∑ j : Fin (n + 1), (a j ^ 2 + 1) := by positivity
+example (a : ℕ → ℤ) : 0 < ∑ j in ({1} : Finset ℕ), (a j ^ 2 + 1) := by
+  have : Finset.Nonempty {1} := singleton_nonempty 1
+  positivity
+example (s : Finset ℕ) : 0 ≤ ∑ j in s, j := by positivity
+example (s : Finset ℕ) : 0 ≤ s.sum id := by positivity
+example (s : Finset ℕ) (f : ℕ → ℕ) (a : ℕ) : 0 ≤ s.sum (f a) := by positivity
```
Comment (eric-wieser) [inline test/positivity.lean:337]: Can you add a test that sums over `Empty` too?

File: Mathlib/Algebra/BigOperators/Order.lean (hunk 5 above)
Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Order.lean:839]: I'm putting it in `Data.Finset.Basic` for now. I will want it available for the `Finset.card` extension.

**Comment by YaelDillies** [comment]
I called the attribute `finsetNonempty` to make sure people don't get confused with the predicate `Finset.Nonempty`. I just tagged plenty of lemmas with the new rule-set. However `proveFinsetNonempty` currently seems to never return a proof. I've copied the two relevant tests back into `Algebra.BigOperators.Order` if you want to have a look.

File: Mathlib/Data/Finset/Basic.lean [hunk 7]
```diff
@@ -3995,3 +3998,30 @@ end List
...
+* `aesop unsafe apply (rule_sets [finsetNonempty])` if they risk directing the search to a blind
+  alley.
+-/
+def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
+    MetaM (Option Q(Finset.Nonempty $s)) := do
+  -- Aesop expects to operate on goals, so we're going to make a new goal.
+  let goal ← Lean.Meta.mkFreshExprMVar q(Finset.Nonempty $s)
+  let mvar := goal.mvarId!
+  -- We want this to be fast, so use only the basic and `Finset.Nonempty` specific rules.
+  let rulesets ← Aesop.Frontend.getGlobalRuleSets #[`builtin, `Finset.Nonempty]
```
Comment (Vierkantor) [inline Mathlib/Data/Finset/Basic.lean:4018]: I suppose you were right that the ruleset name and the predicate name were confusing! This was the ruleset name, not the predicate name:
```suggestion
  let rulesets ← Aesop.Frontend.getGlobalRuleSets #[`builtin, `finsetNonempty]
```

Comment (Vierkantor) [reply Mathlib/Data/Finset/Basic.lean:4018]: To debug this, I did `set_option trace.Tactic.positivity.failure true` and saw:
```
[Tactic.positivity.failure] no such rule set: 'Finset.Nonempty'
      (Use 'declare_aesop_rule_set' to declare rule sets.
       Declared rule sets are not visible in the current file; they only become visible once you import the declaring file.)
```

Comment (YaelDillies) [reply Mathlib/Data/Finset/Basic.lean:4018]: :facepalm:

**Review by Vierkantor — APPROVED** [review]
(I thought I approved this yesterday but apparently the comment never got posted!) Since I wrote the aesop invocation bit, could another maintainer do the final check?

maintainer merge

File: Mathlib/Data/Finset/Basic.lean [hunk 8]
```diff
@@ -4007,3 +4010,30 @@ end List
...
+def proveFinsetNonempty {u : Level} {α : Q(Type u)} (s : Q(Finset $α)) :
+    MetaM (Option Q(Finset.Nonempty $s)) := do
+  -- Aesop expects to operate on goals, so we're going to make a new goal.
+  let goal ← Lean.Meta.mkFreshExprMVar q(Finset.Nonempty $s)
+  let mvar := goal.mvarId!
+  -- We want this to be fast, so use only the basic and `Finset.Nonempty`-specific rules.
+  let rulesets ← Aesop.Frontend.getGlobalRuleSets #[`builtin, `finsetNonempty]
+  -- Fail if the new goal is not closed.
+  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
+  let (remainingGoals, _) ← Aesop.search mvar (.some rules)
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Basic.lean:4033]: Does this fail in the monadic sense, or in the sense of hitting the  `return none` below?

Comment (Vierkantor) [reply Mathlib/Data/Finset/Basic.lean:4033]: This is a monadic failure afaict. So you mean it should be:
```suggestion
  -- Fail if the new goal is not closed.
  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
  let (remainingGoals, _) ← Aesop.search mvar (.some rules) <|> return .none
```

Comment (YaelDillies) [reply Mathlib/Data/Finset/Basic.lean:4033]: I agree it's probably cleaner returning `none`

Comment (YaelDillies) [reply Mathlib/Data/Finset/Basic.lean:4033]: Actually this is incorrect syntax and I'm not sure what the correct one is.

Comment (Vierkantor) [reply Mathlib/Data/Finset/Basic.lean:4033]: Sorry! The following works on my machine:
```suggestion
  -- Fail if the new goal is not closed.
  let rules ← Aesop.mkLocalRuleSet rulesets { terminal := true, generateScript := false }
  let (remainingGoals, _) ←
    try Aesop.search mvar (.some rules)
    catch _ => return none
```

**Comment by Vierkantor** [comment]
@eric-wieser can I ping you to take a look again?

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge
